common/src/main/java/io/netty/util/ResourceLeakDetectorFactory.java

```diff
@@ -0,0 +1,78 @@
+import io.netty.util.internal.SystemPropertyUtil;
+import io.netty.util.internal.logging.InternalLogger;
+import io.netty.util.internal.logging.InternalLoggerFactory;
+
+import java.lang.reflect.Constructor;
+import java.security.AccessController;
+import java.security.PrivilegedAction;
+
+/**
+ * This static factory should be used to load ResourceLeakDetectors as needed
+ */
+public abstract class ResourceLeakDetectorFactory {
+    private static final InternalLogger logger = InternalLoggerFactory.getInstance(ResourceLeakDetectorFactory.class);
+
+    private static volatile ResourceLeakDetectorFactory INSTANCE = new ResourceLeakDetectorFactory() {
+        @Override
+        public <T> ResourceLeakDetector<T> loadResourceLeakDetector(final Class<T> resource) {
+            return AccessController.doPrivileged(new PrivilegedAction<ResourceLeakDetector<T>>() {
+                @Override
+                public ResourceLeakDetector<T> run() {
+                    final String customLeakDetector = SystemPropertyUtil.get("io.netty.customResourceLeakDetector");
+                    try {
+                        if (customLeakDetector != null) {
+                            Class<?> detectorClass = Class.forName(customLeakDetector, true,
+                                    PlatformDependent.getSystemClassLoader());
+
+                            if (ResourceLeakDetector.class.isAssignableFrom(detectorClass)) {
+                                Constructor constructor = detectorClass.getConstructor(Class.class);
+                                ResourceLeakDetector<T> leakDetector =
+                                        (ResourceLeakDetector<T>) constructor.newInstance(resource);
+                                logger.info("Loaded custom ResourceLeakDetector: {}", customLeakDetector);
+                                return leakDetector;
+                            } else {
+                                logger.error("Class {} does not inherit from ResourceLeakDetector.",
+                                        customLeakDetector);
+                            }
+                        }
+                    } catch (Throwable t) {
+                        logger.error("Could not load custom resource leak detector provided: " + customLeakDetector, t);
+                    }
+
+                    ResourceLeakDetector<T> resourceLeakDetector = new ResourceLeakDetector<T>(resource);
+                    logger.info("Loaded default ResourceLeakDetector: {}", resourceLeakDetector);
+                    return resourceLeakDetector;
+                }
+            });
+        }
+    };
+
+    public static ResourceLeakDetectorFactory instance() {
+        return INSTANCE;
+    }
+
+    public static void setResourceLeakDetectorFactory(ResourceLeakDetectorFactory factory) {
+        INSTANCE = factory;
+    }
+
+    public abstract <T> ResourceLeakDetector<T> loadResourceLeakDetector(final Class<T> resource);
```

Comment:
Ah, yes of course. Will do :)

What about the general approach? Is that what you had in mind with the factory?